### DISCUSSION TITLE
Support FaceID/Passcode protection

### DISCUSSION THREAD

**Discussion by rnsc** [xibbon/LaTerminal#6] [Ideas]
I think given the potential sensitivity of the information stored in the application, FaceID/Passcode protection would be a good thing to have.

**Comment by migueldeicaza** [comment]
My only thought about this is that the iPhone already ensures that the user enter a passcode or authenticates with touch and face before they can use the apps on the system.   On the other hand, the "Passwords" app in iOS always requires a secondary face-id verification.

What I am wondering here is whether the FaceID should be triggered on every connection, at launch time or coming back to the background, and lastly whether to have a global setting "Always require FaceID/TouchID" vs the laxer default that i have now.

**Reply by rnsc** [reply]
I must say, I'm asking this because I'm used to having this option available in applications that handle sensitive information. And while I agree that the iPhone itself is protected. If someone steals it while it's unlocked I would be worried that someone would then be able to use sensitive applications.

Regarding the FaceID trigger, I'd argue having it at application level only is sufficient at launch with a recheck after coming out of background as you mentioned.

**Reply by beeradmoore** [reply]
Filed similar issue over [here](https://github.com/migueldeicaza/SwiftTermApp/issues/203) on SwiftTermApp. I think FaceID should be triggered when connecting, maybe when multitasking (maybe if last auth, last time in app was > 5 mins), and when viewing keys. 

It may be too much extra overhead or UI confusion but if users could enable/disable each one of those they can get their own balance of security to ease of use.

**Comment by esteban** [comment]
I'm in favor of "Always require FaceID/TouchID" when enabled, if you already have bank apps with that FaceID/TouchID enabled, pretty much you are used to it so not a big deal I think. Also, disconnecting after locking the phone would be the expected behavior.

**Comment by migueldeicaza** [comment]
Few places where this could be done:

(a) Every time the app is launched required a biometric challenge (like 1Password)
(b) Could be done just at connection time for any access to a key that has a password, and controlled by a global setting
(c) Could be controlled on a key-by-key basis

**Comment by jtamboli** [comment]
From a usability perspective, I’d like something like 1Password where it prompts once on launch, but from a security perspective it would be better if you could use SecAccessControl’s [userPresence](https://developer.apple.com/documentation/security/secaccesscontrolcreateflags/1392879-userpresence) to let the device enforce the requirement. If you used [touchIDAuthenticationAllowableReuseDuration](https://developer.apple.com/documentation/localauthentication/accessing_keychain_items_with_face_id_or_touch_id) you might be able to achieve a similar experience — prompt on launch, but still protect the keys.

**Comment by migueldeicaza** [comment]
This feature has now been implemented, and is available in the AppStore release.
